### PR TITLE
Soft Actor Critic (SAC) Model

### PR DESCRIPTION
**Pull request by blahBlahhhJ**
## What does this PR do?

Update for #596 (issue)
Implementation of SAC model for Reinforcement Learning
pendulum_sac_results

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-bolts/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? [not needed for typos/docs]
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-bolts/blob/master/CHANGELOG.md)?

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by blahBlahhhJ** [comment]
@akihironitta 
Hi! Another RL pull request! This is for Soft Actor Critic which supports continuous action space and off-policy learning. Could you review and give me some feedbacks? About the PEP8 error, I tried both before and after and they lead to different errors so I don't know what to do.
Thanks in advance!

File: pl_bolts/models/rl/common/distributions.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        """
+        z = super().rsample()
+        z_logprob = super().log_prob(z)
+        value = torch.tanh(z)
+        correction = torch.log(self.action_scale * (1 - value ** 2) + 1e-7).sum(1)
+        return self.action_scale * value + self.action_bias, z_logprob - correction
+
+    """
+    Some override methods
+    """
```
Comment (Borda) [inline pl_bolts/models/rl/common/distributions.py:62]:
```suggestion
```

File: pl_bolts/models/rl/sac_model.py [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+        self.agent = SoftActorCriticAgent(self.policy)
+
+        # Hyperparameters
+        self.sync_rate = sync_rate
+        self.gamma = gamma
+        self.batch_size = batch_size
+        self.replay_size = replay_size
+        self.warm_start_size = warm_start_size
+        self.batches_per_epoch = batches_per_epoch
+        self.n_steps = n_steps
```
Comment (Borda) [inline pl_bolts/models/rl/sac_model.py:83]: why not use `save_hyperparameters()`?

File: tests/models/rl/unit/test_sac.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    batch = next(iter(xp_dataloader))
+
+    assert len(batch) == 5
+    assert len(batch[0]) == model.hparams.batch_size
+    assert isinstance(batch, list)
+    assert isinstance(batch[0], Tensor)
+    assert isinstance(batch[1], Tensor)
+    assert isinstance(batch[2], Tensor)
+    assert isinstance(batch[3], Tensor)
+    assert isinstance(batch[4], Tensor)
```
Comment (Borda) [inline tests/models/rl/unit/test_sac.py:60]:
```suggestion
    assert all(isinstance(batch[i], Tensor) for i in range(5))
```

**Review by Borda — APPROVED** [review]
add some more typing if possible

**Review by Borda — COMMENTED** [review]
pls add chlog note

File: pl_bolts/models/rl/sac_model.py [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+
+    def configure_optimizers(self) -> Tuple[Optimizer]:
+        """ Initialize Adam optimizer"""
+        policy_optim = optim.Adam(self.policy.parameters(), self.hparams.policy_learning_rate)
+        q1_optim = optim.Adam(self.q1.parameters(), self.hparams.q_learning_rate)
+        q2_optim = optim.Adam(self.q2.parameters(), self.hparams.q_learning_rate)
+        return policy_optim, q1_optim, q2_optim
+
+    @staticmethod
+    def add_model_specific_args(arg_parser: argparse.ArgumentParser, ) -> argparse.ArgumentParser:
```
Comment (Borda) [inline pl_bolts/models/rl/sac_model.py:348]: Consider using `LightningCLI` see: https://devblog.pytorchlightning.ai/auto-structuring-deep-learning-projects-with-the-lightning-cli-9f40f1ef8b36

**Comment by blahBlahhhJ** [comment]
> pls add chlog note

Done

File: pl_bolts/models/rl/sac_model.py (hunk 4 above)
Comment (blahBlahhhJ) [reply pl_bolts/models/rl/sac_model.py:340]: (Same as A2C PR) Hi, I looked a bit into it, and It seems like in order to add the arguments (add_model_specific_args()) in the lightning module, I'll have to write my own subclass of LightningCLI, which defeats the purpose to use it to simplify the code. Let me know if my understanding is wrong. Or is there an example of using it in other RL algorithms?

**Comment by Borda** [comment]
@blahBlahhhJ mind check why so many tests are failing?

**Comment by blahBlahhhJ** [comment]
> @blahBlahhhJ mind check why so many tests are failing?

@Borda Seems like all the tests fail when I set `self.automatic_optimization = False` with an `AttributeError` saying the attribute cannot be found. I take a look in `PytorchLightningModule` and see that this attribute is managed by a [property setter function](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/core/lightning.py#L247), and the [`__setattr__` function](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L1137) is implemented in `pytorch.nn.Module`. So I'm not sure there's anything I can do to solve any of them. (the tests pass on my local Mac environment which is strange).

**Comment by Borda** [comment]
Thank you @blahBlahhhJ for this nice PR
I have just merged the #598 which brought some collision, could you pls chek them

**Comment by blahBlahhhJ** [comment]
> Thank you @blahBlahhhJ for this nice PR
> I have just merged the #598 which brought some collision, could you pls chek them

@Borda Hi, thanks for merging the other PR! I've fixed all the conflict.

File: pl_bolts/models/rl/sac_model.py (hunk 4 above)
Comment (SeanNaren) [reply pl_bolts/models/rl/sac_model.py]: I think it's fine for now!

**Review by SeanNaren — APPROVED** [review]

**Comment by SeanNaren** [comment]
Thanks @blahBlahhhJ for your patience on this one, bolts has been under maintained so apologies on the time taken to get this merged :)

**Pull request merged**